lang/src/main/kotlin/org/partiql/lang/util/NumberExtensions.kt

```diff
@@ -256,3 +257,302 @@ val Number.isPosInf get() = when (this) {
     is Double -> isInfinite() && this > 0
     else -> false
 }
+
+/**
+ * Returns the given BigDecimal with precision equals to mathContext.precision.
+ *
+ * This is for formatting purpose, all the digit that we are supposedly saying is correct will be shown.
+ */
+private fun BigDecimal.roundToDigits(mathContext: MathContext): BigDecimal {
+    val stripped = this.stripTrailingZeros()
+    val scale = stripped.scale() - stripped.precision() + 1
+    val mantissa = stripped.scaleByPowerOfTen(scale)
+    return if (mantissa.precision() != mathContext.precision) {
+        mantissa.round(mathContext).setScale(mathContext.precision - 1).scaleByPowerOfTen(-scale)
+    } else {
+        stripped.round(mathContext)
+    }
+}
+
+/**
+ * Computes the nth root of a given BigDecimal x.
+ * where x needs to be positive.
+ */
+private fun BigDecimal.intRoot(
+    index: Int,
+    mathContext: MathContext
+): BigDecimal {
+    val operationMC = MathContext(
+        if (mathContext.precision + 2 < 0) Int.MAX_VALUE else mathContext.precision + 2,
+        mathContext.roundingMode
+    )
+
+    val tolerance: BigDecimal = BigDecimal.valueOf(5L).movePointLeft(mathContext.precision)
+
+    // using Newton's method
+    // x_i = ( (r-1) (x_i-1) ^ r + n) / r (x_i-1) ^(r-1)
+    // where n is the number whose rth root we want to compute
+    val r = BigDecimal.valueOf(index.toLong())
+    val rMinusOne = BigDecimal.valueOf(index.toLong() - 1L)
+
+    // The initial approximation is x/index.
+    var res = this.divide(r, mathContext)
+
+    var resPrev: BigDecimal
+    do {
+        // x^(index-1)
+        val xToRMinusOne = res.pow(index - 1, operationMC)
+        // x^index
+        val xToR = res.multiply(xToRMinusOne, operationMC)
+
+        // n + (index-1)*(x^index)
+        val numerator = this.add(rMinusOne.multiply(xToR, operationMC), operationMC)
+
+        // (index*(x^(index-1))
+        val denominator = r.multiply(xToRMinusOne, operationMC)
+
+        // x = (n + (index-1)*(x^index)) / (index*(x^(index-1)))
+        resPrev = res
+
+        res = numerator.divide(denominator, operationMC)
+    } while (res.round(mathContext).subtract(resPrev.round(mathContext)).abs() > tolerance)
+    return res
+}
+
+/**
+ * Computes the square root of a given BigDecimal.
+ * See https://dl.acm.org/doi/pdf/10.1145/214408.214413
+ */
+fun BigDecimal.squareRoot(mathContext: MathContext = MATH_CONTEXT): BigDecimal {
+    if (this.signum() < 0) {
+        throw ArithmeticException("Cannot take root of a negative number")
```

Comment:
What about when `this = 0`?

Reply:
It can be handled via the code logic, but I will add a special case here.